src/circuit/gadget/ecc/chip/add.rs

```diff
@@ -0,0 +1,469 @@
+            ]
+            .iter()
+            .map(|poly| q_add.clone() * poly.clone())
+            .collect()
+        });
+    }
+
+    pub(super) fn assign_region<C: CurveAffine>(
+        &self,
+        p: &EccPoint<C>,
+        q: &EccPoint<C>,
+        offset: usize,
+        region: &mut Region<'_, C::Base>,
+    ) -> Result<EccPoint<C>, Error> {
+        // Enable `q_add` selector
+        self.q_add.enable(region, offset)?;
+
+        // Copy point `p` into `x_p`, `y_p` columns
+        util::assign_and_constrain(region, || "x_p", self.x_p, offset, &p.x, &self.perm)?;
+        util::assign_and_constrain(region, || "y_p", self.y_p, offset, &p.y, &self.perm)?;
+
+        // Copy point `q` into `x_qr`, `y_qr` columns
+        util::assign_and_constrain(region, || "x_q", self.x_qr, offset, &q.x, &self.perm)?;
+        util::assign_and_constrain(region, || "y_q", self.y_qr, offset, &q.y, &self.perm)?;
+
+        let (x_p, y_p) = (p.x.value, p.y.value);
+        let (x_q, y_q) = (q.x.value, q.y.value);
+
+        // inv0(x) evaluates to 0 if x = 0, and 1/x otherwise.
+
+        // Assign α = inv0(x_q - x_p)
+        region.assign_advice(
+            || "α",
+            self.alpha,
+            offset,
+            || {
+                let alpha = x_p.zip(x_q).map(|(x_p, x_q)| inv0(x_q - x_p));
+                alpha.ok_or(Error::SynthesisError)
+            },
+        )?;
+
+        // Assign β = inv0(x_p)
+        region.assign_advice(
+            || "β",
+            self.beta,
+            offset,
+            || {
+                let beta = x_p.map(inv0);
+                beta.ok_or(Error::SynthesisError)
+            },
+        )?;
+
+        // Assign γ = inv0(x_q)
+        region.assign_advice(
+            || "γ",
+            self.gamma,
+            offset,
+            || {
+                let gamma = x_q.map(inv0);
+                gamma.ok_or(Error::SynthesisError)
+            },
+        )?;
+
+        // Assign δ = inv0(y_q + y_p) if x_q = x_p, 0 otherwise
+        region.assign_advice(
+            || "δ",
+            self.delta,
+            offset,
+            || {
+                let x_p = x_p.ok_or(Error::SynthesisError)?;
+                let x_q = x_q.ok_or(Error::SynthesisError)?;
+                let y_p = y_p.ok_or(Error::SynthesisError)?;
+                let y_q = y_q.ok_or(Error::SynthesisError)?;
+
+                let delta = if x_q == x_p {
+                    inv0(y_q + y_p)
+                } else {
+                    C::Base::zero()
+                };
+                Ok(delta)
+            },
+        )?;
+
+        #[allow(clippy::collapsible_else_if)]
+        // Assign lambda
+        let lambda = x_p
+            .zip(y_p)
+            .zip(x_q)
+            .zip(y_q)
+            .map(|(((x_p, y_p), x_q), y_q)| {
+                if x_q != x_p {
+                    // λ = (y_q - y_p)/(x_q - x_p)
+                    (y_q - y_p) * (x_q - x_p).invert().unwrap()
```

Comment:
`inv0(x_q - x_p)`, which we compute above for `alpha`, would suffice here since we know that `x_q != x_p` in this branch.